### PR TITLE
revise df-sb

### PR DESCRIPTION
**Pull request by icecream17**
df-ssb imported from BJ's mathbox

This removes axioms in many cases (although adds axioms in some cases)

I don't actually know the reason for the [y/x]ph to [t/x]ph change but I'm keeping it for now

Eventually df-sbOLD will become a theorem in another commit, but the proof for such a theorem would use many axioms (mp-7 and 10-13 it seems) <https://github.com/metamath/set.mm/pull/3214#issuecomment-1566277538>.

So the plan is to import theorems from bj's mathbox, then change any theorems still using df-sbOLD if I can find a better proof, then using dfsb__ for the rest.

---------
Done: sb1 sb2 sb4 sb4a hbsb2 nfsb2 equsb1 sb6f sb5f nfsb4t nfsb4 sbi1 sbim sbrim sban sbbi sblbis sbie sbied
Proving dfsb11: sbco2 sb7f

### CODE REVIEW

File: set.mm
Comment (tirix) [inline set.mm:15798]: "Once it is proved" refers to the definition like it was a theorem. Maybe rather "With this definition, ..."

File: set.mm
Comment (tirix) [inline set.mm:15932]: Would it be possible to prove this as a theorem (e.g. `dfsb2`), and then replace all instances of `df-sb` with `dfsb2`?
These would still use `ax-13`, but then it's possible to improve them later.

File: changes-set.txt
Comment (jkingdon) [inline changes-set.txt:30]: Should all these "merged" say "moved"? I'm not sure what the word "merged" would mean here.

File: set.mm
Comment (jkingdon) [reply set.mm:15932]: I've used versions of this approach with good results in the past. For example in changing the definitions of `seq` and `sum_` in iset.mm (this was spread over many pull requests from 2022 to 2023).

I can see two variations:
1. Change the definition (as in this pull request), then prove `df-sbOLD` (or whatever we call it) and change existing usages, as @tirix suggests in the comment above
2. Don't change the definition yet, but introduce our desired new definition as a theorem (`dfsb7al` or whatever we'd call it) and slowly change usages from `df-sb` to the new one. At some later point we'd change it from a theorem to a definition.

What is in the pull request now is giving an error `Axiom df-sbOLD has failed the definitional soundness check. The previous axiom df-sb uses the symbol being defined in this axiom.` and I don't think we'd want to relax the definition check given the potential for introducing unsoundness if we missed anything about whether the definitions are equivalent.

File: changes-set.txt
Comment (icecream17) [reply changes-set.txt:34]: It means that the proof already existed, so instead of copy and pasting, the two theorems were "merged" together. I just made this up so if "moved" would still be better then I'll change it

Comment (jkingdon) [reply changes-set.txt:34]: Maybe something like "bj-ssbn was an alternative proof of the same theorem"? I'm not sure it is necessary to mention the mathbox part but neither "moved" nor "merged" really explains what happened, at least to my eyes.

**Comment by benjub** [comment]
This implements #3003

File: set.mm
Comment (icecream17) [reply set.mm]: I think `this` was referring to `sbjust`, at which point there would be axioms mp thru 7.

I ended up deleting the paragraph though (https://github.com/metamath/set.mm/pull/3283/commits/d66989773f1fa16a5df1312b481040ba3db8aa5d)

Comment (benjub) [reply set.mm]: Indeed, that sentence was meant to be temporary, and deleted once the change has been made.

File: changes-set.txt
Comment (benjub) [reply changes-set.txt:34]: I'm fine with either, or even not mentioning it in `changes-set.txt`.

Comment (icecream17) [reply changes-set.txt]: the wording was revised in https://github.com/metamath/set.mm/pull/3283/commits/c104a28a036c8e3adbac3c93a481ab97ae94e14a

Comment (jkingdon) [reply changes-set.txt]: New wording looks good to me

**Comment by wlammen** [comment]
This cannot be merged before the conflicts are resolved.

**Comment by tirix** [comment]
> This cannot be merged before the conflicts are resolved.

I did the merge and then continued with the necessary fixes for this PR to be accepted (rewrap and discouraged markers).
@icecream17 I hope you don't mind - you'll have to pull those changes first if you want to do further edits to this PR.

**Comment by wlammen** [comment]
The CI process sometimes chokes on loading the LATEX libraries.  A few weeks ago I suffered from the same problem.  Nothing helped.  I repeatedly entered the "Checks" tab and requested a "Re-run failed jobs" (drop-down button on the upper right corner), and hoped for the best (a few hours without success )-: ).

**Comment by wlammen** [comment]
In total, the new version of substitution leads to a smaller axiom footprint of sb related theorems, the obvious exemptions are the direct applications of the old definition.  Nice work!  Kudos to all that contributed to this convincing result.

**Comment by david-a-wheeler** [comment]
I forced the rerun of the LaTeX job. In the long term we'll need to find a lighter way to load the LaTeX packages, but I don't want to hold up this great work just for that.

File: set.mm
Comment (benjub) [inline set.mm:15754]:
```suggestion
       For example, ` [ x / y ] z e. y ` is the same as ` z e. x ` (when ` y ` and ` z ` are distinct), as shown
```

File: set.mm
Comment (benjub) [inline set.mm:15753]: The statement defines ` [ t / x ] ` with dummy variable ` y ` while this part of the comment uses ` [ y / x ] ` . Either the comment or the statement has to be changed. I had used ` [ t / x ] ` and dummy ` y ` to underline that x, y are bound while t is free (t is like a "term"), but maybe it's better to follow the "mathematical alphabet": x, y, z...

File: set.mm
Comment (benjub) [inline set.mm:15757]: Not very clear.

File: set.mm
Comment (benjub) [inline set.mm:15768]: A very similar notation, namely ` ( y | x ) ph ` was introduced in Bourbaki's Set Theory (Chapter 1, Description of Formal Mathematic, 1953).

(for the curious: http://archives-bourbaki.ahp-numerique.fr/files/original/9f495d5e626460cc111fb280df8227c1.pdf)

File: set.mm
Comment (benjub) [inline set.mm:15795]: Maybe something like:
```
...1993.)  Revised from the original definition ~ dfsb1 .  (Revised...
```
to give a very brief explanation.

File: set.mm
Comment (benjub) [inline set.mm:16511]:
```suggestion
    $( Version of ~ sbcom3 with a disjoint variable condition using fewer axioms.
```

File: set.mm
Comment (benjub) [inline set.mm:16914]: sbjust should be right before df-sb, as is usual for justification theorems

File: set.mm
Comment (benjub) [inline set.mm:20939]: Maybe add the sentences: "This was the original definition before ~ df-sb. Note that it does not require dummy variables."

**Comment by benjub** [comment]
I just made a few comments above. After this is merged, there will probably be a few more passes to harvest additional benefits of the definition change and still reduce axiom usages and proof lengths (including through some reorderings of results). Thanks a lot @icecream17 for this work, and thanks also to the reviewers.

**Comment by david-a-wheeler** [comment]
Ugh. The metamath-exe job can't install some of the TeX packages, with these messages:

~~~~
Err:4 http://azure.archive.ubuntu.com/ubuntu jammy-updates/main amd64 libgs9-common all 9.55.0~dfsg1-0ubuntu5.2
  404  Not Found [IP: 40.119.46.219 80]
Err:8 http://azure.archive.ubuntu.com/ubuntu jammy-updates/main amd64 libgs9 amd64 9.55.0~dfsg1-0ubuntu5.2
  404  Not Found [IP: 40.119.46.219 80]
Err:52 http://azure.archive.ubuntu.com/ubuntu jammy-updates/main amd64 ghostscript amd64 9.55.0~dfsg1-0ubuntu5.2
  404  Not Found [IP: 40.119.46.219 80]
E: Failed to fetch http://azure.archive.ubuntu.com/ubuntu/pool/main/g/ghostscript/libgs9-common_9.55.0%7edfsg1-0ubuntu5.2_all.deb  404  Not Found [IP: 40.119.46.219 80]
Fetched 838 MB in 37s (22.8 MB/s)
E: Failed to fetch http://azure.archive.ubuntu.com/ubuntu/pool/main/g/ghostscript/libgs9_9.55.0%7edfsg1-0ubuntu5.2_amd64.deb  404  Not Found [IP: 40.119.46.219 80]
E: Failed to fetch http://azure.archive.ubuntu.com/ubuntu/pool/main/g/ghostscript/ghostscript_9.55.0%7edfsg1-0ubuntu5.2_amd64.deb  404  Not Found [IP: 40.119.46.219 80]
E: Unable to fetch some archives, maybe run apt-get update or try with --fix-missing?
Error: Process completed with exit code 100.
~~~~

The obvious solution is to try to run `apt-get update --fix-missing` before it.

**Comment by david-a-wheeler** [comment]
Here's my plan: I'm going to try to create a branch that fixes the CI/CD pipeline. If that works, you can merge the "develop" branch into this branch (so we know it all works), then merge this branch back into develop. Sorry for the CI/CD problems.

**Comment by david-a-wheeler** [comment]
@icecream17 - can you do a "git pull" and merge from the metamath develop branch, then re-push? I *think* that will solve the CI/CD problem this proposed change is having. I'm sorry for the trouble, and I'm really pleased to see this work - I want to see it merged in!

**Comment by david-a-wheeler** [comment]
@wlammen - I'm not *certain* what the LaTeX problems are. My current hypothesis is that because we didn't do an update, we were trying to pull really old packages that aren't in many caches, & so the fetches were timing out. I've snuck in an apt-get update first, so that we're getting current dependencies, with the *hope* that this resolves the problem.

**Comment by wlammen** [comment]
When I had this (or a similar) Latex problem, they vanished after some time. The 404 error indicates the service at http://azure.archive.ubuntu.com/ubuntu/pool/main/g/ghostscript is not available, but since it was a few hours ago, this may well be a temporary external problem.  Maybe they update the sources right now, and during this process the server is shut down?

**Comment by tirix** [comment]
Could it be that the Azure servers are receiving too many requests from GitHub, assume it's a kind of attack, and decide to ignore them for a while? If that's the case, maybe some kind of caching the required packages would help? (wild guesses here)

**Comment by david-a-wheeler** [comment]
The job is now failing correctly :-).

It appears that a statement comment incorrectly has "df-sb." instead of "df-sb" - can someone track down the extra period after "df-sb" in a comment somewhere?

Here's the error response:

~~~~
Run scripts/verify --top_date_skip --extra 'write bibliography mmbiblio.html' set.mm
Metamath - Version 0.198 7-Aug-2021           Type HELP for help, EXIT to exit.
MM> !SET ECHO ON
Command line echoing is now turned on.
MM> !SET SCROLL CONTINUOUS
Continuous scrolling is now in effect.
MM> !READ set.mm
Reading source file "set.mm"... 43254886 bytes
43254886 bytes were read into the source buffer.
The source has 205948 statements; 2746 are $a and 40959 are $p.
No errors were found.  However, proofs were not checked.  Type VERIFY PROOF *
if you want to check them.
MM> !VERIFY PROOF *
0 10%  20%  30%  40%  50%  60%  70%  80%  90% 100%
..................................................
All proofs in the database were verified in 5.58 s.
MM> !VERIFY MARKUP * / TOP_DATE_SKIP
Checking statement label conventions...
Checking latexdef, htmldef, althtmldef...
Checking statement comments...
?Warning: The label token "df-sb." (referenced in comment of statement "sbbid")
is not a $a or $p statement label.
Checking section header comments...
Checking bibliographic references...
?Sorry, couldn't open the file "mmbiblio.html".
Checking mathbox independence...
MM> The old file will be renamed mmbiblio.html~1.
!WRITE BIBLIOGRAPHY mmbiblio.html
?Sorry, couldn't open the file "mmbiblio.html".
MM> !QUIT
Error: Process completed with exit code 1.
~~~~

**Comment by david-a-wheeler** [comment]
The problem is that there is a reference to "df-sb." in the comment of statement "sbbid" - presumably there should be a space just before the period.

@icecream17 - can you fix the sbbid comment?

File: set.mm
Comment (icecream17) [reply set.mm]: Based on <https://math.stackexchange.com/questions/2659471/what-does-it-mean-for-a-variable-to-be-free-for-something>, a "free for" condition ensures that a substitution is valid.

It's probably the "somewhat complicated recursive definition with multiple cases based on the occurrences of free and bound variables in [a] wff".

This seems kinda tautological so I deleted it and mentioned "free for" in the second paragraph
https://github.com/metamath/set.mm/pull/3283/commits/29d849be34c55be7bdb2110223e33e0395daedcf

**Review by wlammen — APPROVED** [review]

**Comment by david-a-wheeler** [comment]
Whew! Looks like we have a version that is passing all tests and ready to merge. Thanks for your patience.

File: set.mm
Comment (benjub) [reply set.mm]: Thank you, it was better to remove it indeed: I wrote "not clear", but I meant and should have written "nonsensical". The link you make with the stackexchange question is not clear either. I would remove that link since better explanations can be found in any standard logic textbook.

As for the "somewhat complicated...": it refers to the recursive definitions of free and bound variables in standard first-order logic (and these notions are necessary to define proper substitution).

**Comment by wlammen** [comment]
I'd rather see somebody else check this in, to get a feeling of consent.  But waiting longer is no real option either, in my eyes.  Whatever needs to be done from here, can be done in a separate pull request.

**Pull request merged**
